### DISCUSSION TITLE
DISTINCT ON doesn't work for common table expressions in postgresql

### DISCUSSION THREAD

**Discussion by slymon99** [sqlalchemy/sqlalchemy#7899] [Usage Questions]
### Describe the bug

Creating a query with DISTINCT ON works for a regular query, but fails to work when the query is turned into a subquery. A warning is also printed. It seems like when you convert the query to a CTE, sqlalchemy doesn't realize it's still in Postgres in terms of DISTINCT ON support.

### To Reproduce

```python
> str(session.query(Fruit).distinct(Fruit.name))
out: SELECT DISTINCT ON (fruits.name) fruits.color AS ...
> str(session.query(Fruit).distinct(Fruit.name).cte())
SADeprecationWarning: DISTINCT ON is currently supported only by the PostgreSQL dialect.  Use of DISTINCT ON for other backends is currently silently ignored, however this usage is deprecated, and will raise CompileError in a future release for all backends that do not support this syntax.
out: SELECT DISTINCT fruits.color AS ...
```

### Error

SADeprecationWarning: DISTINCT ON is currently supported only by the PostgreSQL dialect.  Use of DISTINCT ON for other backends is currently silently ignored, however this usage is deprecated, and will raise CompileError in a future release for all backends that do not support this syntax.

### Versions

- OS: Ubuntu 18.04.5
- Python: 3.9.5
- SQLAlchemy: 1.4.34
- Database: postgres
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pyscopg

**Comment by CaselIT** [comment]
Hi,

That's expected. using `str()` on a query it will invoke the default string compiler that knows nothing about postgresql.
Please refer to https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#faq-sql-expression-string that shows how to compile a statement for a particular database
